.github/workflows/sync-opencraft-forks-with-upstream.yml

```diff
@@ -0,0 +1,86 @@
+            upstream_branch: open-release/palm.master
+
+    runs-on: ubuntu-latest
+    env:
+      # TODO: set this to the OpenCraft Requirements Bot PAT
+      GITHUB_TOKEN: ${{ secrets.GITHUB_BOT_TOKEN }}
+
+    steps:
+      - name: Get and store current datetime stamp
+        id: current-datetime-stamp
+        run: echo "DATE_TIME_STAMP=$(date +'%Y%m%d-%s')" >> "$GITHUB_OUTPUT"
+
+      - name: Construct and store sync branch name
+        id: sync-branch
+        run: echo "SYNC_BRANCH=sync-${{ matrix.upstream_branch }}-${{ steps.current-datetime-stamp.outputs.DATE_TIME_STAMP }}" >> "$GITHUB_OUTPUT"
+
+      - name: Configure git to use GitHub CLI for credentials
+        run: gh auth setup-git -h github.com
+
+      - name: Checkout Forked Repo
+        uses: actions/checkout@v3
+        with:
+          repository: ${{ matrix.fork_repo }}
+          token: ${{ secrets.GITHUB_BOT_TOKEN }}
+
+      - name: Add Remote Upstream
+        run: git remote add upstream ${{ matrix.upstream_repo }}
+
+      - name: Fetch Latest from Upstream
+        run: git fetch upstream
+
+      - name: Create a new branch off Upstream master
+        run: git checkout -b ${{ steps.sync-branch.outputs.SYNC_BRANCH }} upstream/${{ matrix.upstream_branch }}
+
+      - name: Check for changes Upstream
+        id: check-sync-required
+        run: echo "SYNC_REQUIRED=$(git log ${{ steps.sync-branch.outputs.SYNC_BRANCH }} ^${{ matrix.fork_branch }} | grep -q . && echo true)" >> "$GITHUB_OUTPUT"
+
+      - name: Push new branch to Fork
+        if: ${{ steps.check-sync-required.outputs.SYNC_REQUIRED }}
+        run: git push origin
+
+      - name: Create PR on Fork with latest changes from Upstream
+        if: ${{ steps.check-sync-required.outputs.SYNC_REQUIRED }}
+        run: >
+          gh pr create
+          --repo ${{ matrix.fork_repo }}
+          --base ${{ matrix.fork_branch }}
+          --head $(git branch --show-current)
+          --title "Sync ${{ matrix.fork_branch }} with Upstream ${{ steps.current-datetime-stamp.outputs.DATE_TIME_STAMP }}"
+          --body '
+          Syncing ${{ matrix.fork_branch }} with Upstream
+
+          ## Note on Conflicts ⚠️
+
+          In cases of conflicts you can go ahead and resolve it here on Github if it is simple enough.
+          However if it is a more complicated conflict please follow the steps below:
+
+            1. Check out `${{ steps.sync-branch.outputs.SYNC_BRANCH }}` locally:
+            2. Pull latest changes from `${{ matrix.fork_branch }}` into that branch, make sure your `[REMOTE]` is pointing to this repo:
+            ```sh
+            git pull [REMOTE] ${{ matrix.fork_branch }}
+            ```
+            3. Resolve the conflicts locally then commit the result, this will create a new merge commit
+            4. Push the new merge commit to `${{ steps.sync-branch.outputs.SYNC_BRANCH }}` to update this PR
+            5. Review the PR again and merge when ready!'
```

Comment:
So I spent some time figuring out the differences between the 3 options and how it will affect us. I do want to note that I had to change my implementation slightly since, for some reason, there were discrepancies from running the various git commands to check for differences with upstream when run locally on my machine and when running inside the Github Action. I tried the following:

- `git log [sync_branch] ^[fork_branch]`
- `git diff [fork_branch]..` (running it when on the [sync_branch])
- `git merge-base --is-ancestor [sync_branch] [fork_branch]`
- `git cherry [fork_branch] [sync_branch]`

In all of these commands I was able to determine if there were changes upstream when I ran them locally, however when it ran part of the Github Action, it just wouldn't give me the same results. So I found it was more reliable to just let the `gh pr create ...` command take care of it. It fails if there are no commit between the sync_branch and fork_branch, then I would just handle it and delete the sync branch. This approach worked much more reliably.

Now back to the 3 different merge approaches, here are my findings:
1. **Create a merge commit:** This one is the most straightforward option, although it adds merge commits to our fork, it makes things easier from for resolving conflicts and merging into the fork. Even when the Github Action runs again, it can more reliably detect if there were new changes from upstream that need to synced or not.
2. **Squash and merge:** This one caused issues when I ran the Github Action again after the merge, a new PR would be created with the same changes that were merged earlier, just different commit.
3. **Rebase and merge:** This option was not allowed when there were conflicts in the PR, even after resolving them. I figured we probably don't want this any way since like you mentioned, might cause us some headaches with commit history and other peoples' branches if we try rebasing to the main fork branch.

I went with the first option, "Create a merge commit", and added a note to the PR comment to let the reviewer know to use that when merging the PR. Since it seems like it fits our use case the most.

Reply:
@yusuf-musleh, thank you for the detailed investigation. Sounds good!